lta/site_move_verifier.py

```diff
@@ -0,0 +1,172 @@
+        lta_rc = RestClient(self.lta_rest_url,
+                            token=self.lta_rest_token,
+                            timeout=self.work_timeout_seconds,
+                            retries=self.work_retries)
+        self.logger.info("Asking the LTA DB for a Bundle to verify.")
+        pop_body = {
+            "claimant": f"{self.name}-{self.instance_uuid}"
+        }
+        response = await lta_rc.request('POST', f'/Bundles/actions/pop?dest={self.dest_site}&status=transferring', pop_body)
+        self.logger.info(f"LTA DB responded with: {response}")
+        bundle = response["bundle"]
+        if not bundle:
+            self.logger.info("LTA DB did not provide a Bundle to verify. Going on vacation.")
+            return False
+        # process the Bundle that we were given
+        await self._verify_bundle(lta_rc, bundle)
+        return True
+
+    async def _verify_bundle(self, lta_rc: RestClient, bundle: BundleType) -> bool:
+        """Verify the provided Bundle with the transfer service and update the LTA DB."""
+        bundle_id = bundle["uuid"]
+        # instantiate a TransferService to query about the bundle
+        xfer_service = instantiate(self.transfer_config)
+        # ask the transfer service about the status of the transfer
+        xfer_status = await xfer_service.status(bundle["transfer_reference"])
+        self.logger.info(f"Bundle transfer status is: {xfer_status}")
+        # if it's not completed, we're still waiting to verify it
+        if not xfer_status["completed"]:
+            self.logger.info(f"Transfer of Bundle {bundle_id} is incomplete and will not be verified at this time.")
+            return False
+        # when we verify bundles, we do so at the destination site
+        dest = bundle["dest"]
+        pfn_prefix = self.transfer_config["sites"][dest]["pfn"]  # UGLY: hard-coded RucioTransferService dependency
+        parsed_url = urlparse(pfn_prefix)
+        rucio_path = parsed_url.path
+        bundle_name = os.path.basename(bundle["bundle_path"])
+        bundle_path = os.path.join(rucio_path, bundle_name)
+        # we'll compute the bundle's checksum
+        self.logger.info(f"Computing SHA512 checksum for bundle: '{bundle_path}'")
+        checksum_sha512 = sha512sum(bundle_path)
+        self.logger.info(f"Bundle '{bundle_path}' has SHA512 checksum '{checksum_sha512}'")
+        # now we'll compare the bundle's checksum
+        if bundle["checksum"]["sha512"] != checksum_sha512:
+            self.logger.info(f"SHA512 checksum at the time of bundle creation: {bundle['checksum']['sha512']}")
+            self.logger.info(f"SHA512 checksum of the file at the destination: {checksum_sha512}")
+            self.logger.info(f"These checksums do NOT match, and the Bundle will NOT be verified.")
+            bundle["status"] = "quarantined"
+            bundle["reason"] = f"Checksum mismatch between creation and destination: {checksum_sha512}"
```

Comment:
We could delete the local file and attempt to prod Rucio into transferring it again.

In practice, I imagine it will play out more like this:
- If we've already exceeded our retry limit, execute the code as written now, otherwise:
- Log the problematic checksum mismatch
- Delete the local copy
- Update the retry count in the Bundle record in the LTA DB
- Wait for the Rucio deamon to check the destination RSE and start another copy

Honestly, I'd rather avoid the retry code until we know how common such problems are and why they occur.
If Rucio is telling us a replica is AVAILABLE and the sha512 checksums don't match, that gives me *serious* concerns about the reliability of Rucio. And until I know why Rucio might get into that state, I'd rather such a case show up fast/loud and catch my attention immediately rather than get hidden behind a retry mechanism.

Reply:
That's true.  Forgot Rucio has some internal checks.